### ISSUE TITLE
A confusing question in transformer training

### ISSUE THREAD

**Issue by forever-rz** [liuqk3/PUT#15]
Thanks for your contribution,but there is a problem when i train it on FFHQ.Once the ratio of mask is larger, it seems that only part of the completed result is repaired, the part that is not repaired stays black and no new content seems to be generated. Is this normal?
eg1:  First on the left, second on the left remain some strange black regions .(epoch:12)
mask_input 
![image](https://user-images.githubusercontent.com/60658086/200735972-ba2c192b-fdb6-4f7c-8009-993b57dfbc7b.png)
completed
![image](https://user-images.githubusercontent.com/60658086/200736012-ed1565b1-639f-46e5-aea0-efabf44d8b95.png)
reconstruction
![image](https://user-images.githubusercontent.com/60658086/200736069-58272df7-92fe-4ce2-8944-78215ed6f556.png)

**Comment by forever-rz** [comment]
This bad visual effect became increasingly apparent later in the training.
eg2:   epoch:19
mask_input
![image](https://user-images.githubusercontent.com/60658086/200738888-426e3353-5f28-4626-8d6e-581afbda828d.png)
completed
![image](https://user-images.githubusercontent.com/60658086/200738926-b77d4541-5c2a-4e64-b424-d598e5c8b7ca.png)
reconstruction
![image](https://user-images.githubusercontent.com/60658086/200739160-2ec81a95-c492-4eef-a9c4-ad94027918a0.png)
input
![image](https://user-images.githubusercontent.com/60658086/200739269-61b7659f-7c4f-446d-822f-8ded1218c9fa.png)

**Comment by liuqk3** [comment]
Hi @forever-rz , Thanks for your interests.
It seems that the second codebook is used for quantization while training transformer. For FFHQ, it will not take so long to get a reasonable inpainting results.

**Comment by forever-rz** [comment]
> Hi @forever-rz , Thanks for your interests. It seems that the second codebook is used for quantization while training transformer. For FFHQ, it will not take so long to get a reasonable inpainting results.
Thanks for your help, Can you tell me how should I set up just one codebook then? the transformer training configs is below:
dataloader:
  batch_size: 16
  data_root: data
  num_workers: 1
  train_datasets:
  - params:
      im_preprocessor_config:
        params:
          horizon_flip: true
          random_crop: true
          size:
          - 256
          - 256
          smallest_max_size: 256
        target: image_synthesis.data.utils.image_preprocessor.SimplePreprocessor
      image_list_file: data/ffhqtrain_69k.txt
      mask: 1.0
      mask_low_size:
      - 32
      - 32
      mask_low_to_high: 0.1
      multi_image_mask: false
      name: ffhq
      provided_mask_name: irregular-mask/testing_mask_dataset
      return_data_keys:
      - image
      - mask
      stroken_mask_params:
        keep_ratio:
        - 0.3
        - 0.6
        maxBrushWidth: 30
        maxLength: 100
        maxVertex: 10
        minBrushWidth: 10
        minVertex: 5
        min_area: 64
      use_provided_mask: 0.8
      use_provided_mask_ratio:
      - 0.3333333
      - 1.0
    target: image_synthesis.data.image_list_dataset.ImageListDataset
  validation_datasets:
  - params:
      # data_root: /home/liuqk/Program/python/image-synthesis/data
      im_preprocessor_config:
        params:
          size:
          - 256
          - 256
          smallest_max_size: 256
        target: image_synthesis.data.utils.image_preprocessor.SimplePreprocessor
      image_list_file: data/ffhqvalidation_1k.txt
      mask: 1.0
      mask_low_size:
      - 32
      - 32
      mask_low_to_high: 0.0
      multi_image_mask: false
      name: ffhq
      provided_mask_name: irregular-mask/testing_mask_dataset
      return_data_keys:
      - image
      - mask
      - relative_path
      stroken_mask_params:
        keep_ratio:
        - 0.3
        - 0.6
        maxBrushWidth: 30
        maxLength: 100
        maxVertex: 10
        minBrushWidth: 10
        minVertex: 5
        min_area: 64
      use_provided_mask: 1.0 #TODO0.8
      use_provided_mask_ratio:
      - 0.4
      - 1.0
    target: image_synthesis.data.image_list_dataset.ImageListDataset

model:
  target: image_synthesis.modeling.models.masked_image_inpainting_transformer_in_feature.MaskedImageInpaintingTransformer
  params:
    n_layer: 30
    content_seq_len: 1024
    n_embd: 512
    n_head: 8

    num_token: 512

    embd_pdrop: 0.0
    attn_pdrop: 0.0
    resid_pdrop: 0.0
    attn_content_with_mask: False
    mlp_hidden_times: 4
    block_activate: GELU2
    random_quantize: 0.3
    weight_decay: 0.01

    content_codec_config:
      target: image_synthesis.modeling.codecs.image_codec.patch_vqgan.PatchVQGAN
      params:
        ckpt_path: OUTPUT/pvqvae_ffhq/checkpoint/last.pth
        trainable: False
        token_shape: [32, 32]
        combine_rec_and_gt: True
        quantizer_config:
          target: image_synthesis.modeling.codecs.image_codec.patch_vqgan.VectorQuantizer
          params:
            n_e: 1024
            e_dim: 256
            masked_embed_start: 512
            embed_ema: True
            get_embed_type: retrive
            distance_type: euclidean
        encoder_config:
          target: image_synthesis.modeling.codecs.image_codec.patch_vqgan.PatchEncoder2
          params:
            in_ch: 3
            res_ch: 256
            out_ch: 256
            num_res_block: 8
            res_block_bottleneck: 2
            stride: 8
        decoder_config:
          target: image_synthesis.modeling.codecs.image_codec.patch_vqgan.PatchConvDecoder2
          params:
            in_ch: 256
            out_ch: 3
            res_ch: 256
            num_res_block: 8
            res_block_bottleneck: 2
            stride: 8
            up_layer_with_image: true
            encoder_downsample_layer: conv

solver:
  adjust_lr: none
  base_lr: 0.0
  find_unused_parameters: false
  max_epochs: 250
  optimizers_and_schedulers:
  - name: transformer
    optimizer:
      params:
        betas: !!python/tuple
        - 0.9
        - 0.95
      target: torch.optim.AdamW
    scheduler:
      params:
        min_lr: 1.0e-05
        warmup: 2000
        warmup_lr: 0.0003
      step_iteration: 1
      target: image_synthesis.engine.lr_scheduler.CosineAnnealingLRWithWarmup
  sample_iterations: 400
  save_epochs: 1
  validation_epochs: 1

**Comment by forever-rz** [comment]
dataloader:
  batch_size: 16
  data_root: data
  num_workers: 1
  train_datasets:
  - params:
      im_preprocessor_config:
        params:
          horizon_flip: true
          random_crop: true
          size:
          - 256
          - 256
          smallest_max_size: 256
        target: image_synthesis.data.utils.image_preprocessor.SimplePreprocessor
      image_list_file: data/ffhqtrain_69k.txt
      mask: 1.0
      mask_low_size:
      - 32
      - 32
      mask_low_to_high: 0.1
      multi_image_mask: false
      name: ffhq
      provided_mask_name: irregular-mask/testing_mask_dataset
      return_data_keys:
      - image
      - mask
      stroken_mask_params:
        keep_ratio:
        - 0.3
        - 0.6
        maxBrushWidth: 30
        maxLength: 100
        maxVertex: 10
        minBrushWidth: 10
        minVertex: 5
        min_area: 64
      use_provided_mask: 0.8
      use_provided_mask_ratio:
      - 0.3333333
      - 1.0
    target: image_synthesis.data.image_list_dataset.ImageListDataset
  validation_datasets:
  - params:
      # data_root: /home/liuqk/Program/python/image-synthesis/data
      im_preprocessor_config:
        params:
          size:
          - 256
          - 256
          smallest_max_size: 256
        target: image_synthesis.data.utils.image_preprocessor.SimplePreprocessor
      image_list_file: data/ffhqvalidation_1k.txt
      mask: 1.0
      mask_low_size:
      - 32
      - 32
      mask_low_to_high: 0.0
      multi_image_mask: false
      name: ffhq
      provided_mask_name: irregular-mask/testing_mask_dataset
      return_data_keys:
      - image
      - mask
      - relative_path
      stroken_mask_params:
        keep_ratio:
        - 0.3
        - 0.6
        maxBrushWidth: 30
        maxLength: 100
        maxVertex: 10
        minBrushWidth: 10
        minVertex: 5
        min_area: 64
      use_provided_mask: 1.0 #TODO0.8
      use_provided_mask_ratio:
      - 0.4
      - 1.0
    target: image_synthesis.data.image_list_dataset.ImageListDataset

model:
  target: image_synthesis.modeling.models.masked_image_inpainting_transformer_in_feature.MaskedImageInpaintingTransformer
  params:
    n_layer: 30
    content_seq_len: 1024
    n_embd: 512
    n_head: 8

    num_token: 512

    embd_pdrop: 0.0
    attn_pdrop: 0.0
    resid_pdrop: 0.0
    attn_content_with_mask: False
    mlp_hidden_times: 4
    block_activate: GELU2
    random_quantize: 0.3
    weight_decay: 0.01

    content_codec_config:
      target: image_synthesis.modeling.codecs.image_codec.patch_vqgan.PatchVQGAN
      params:
        ckpt_path: OUTPUT/pvqvae_ffhq/checkpoint/last.pth
        trainable: False
        token_shape: [32, 32]
        combine_rec_and_gt: True
        quantizer_config:
          target: image_synthesis.modeling.codecs.image_codec.patch_vqgan.VectorQuantizer
          params:
            n_e: 1024
            e_dim: 256
            masked_embed_start: 512
            embed_ema: True
            get_embed_type: retrive
            distance_type: euclidean
        encoder_config:
          target: image_synthesis.modeling.codecs.image_codec.patch_vqgan.PatchEncoder2
          params:
            in_ch: 3
            res_ch: 256
            out_ch: 256
            num_res_block: 8
            res_block_bottleneck: 2
            stride: 8
        decoder_config:
          target: image_synthesis.modeling.codecs.image_codec.patch_vqgan.PatchConvDecoder2
          params:
            in_ch: 256
            out_ch: 3
            res_ch: 256
            num_res_block: 8
            res_block_bottleneck: 2
            stride: 8
            up_layer_with_image: true
            encoder_downsample_layer: conv

solver:
  adjust_lr: none
  base_lr: 0.0
  find_unused_parameters: false
  max_epochs: 250
  optimizers_and_schedulers:
  - name: transformer
    optimizer:
      params:
        betas: !!python/tuple
        - 0.9
        - 0.95
      target: torch.optim.AdamW
    scheduler:
      params:
        min_lr: 1.0e-05
        warmup: 2000
        warmup_lr: 0.0003
      step_iteration: 1
      target: image_synthesis.engine.lr_scheduler.CosineAnnealingLRWithWarmup
  sample_iterations: 400
  save_epochs: 1
  validation_epochs: 1

**Comment by forever-rz** [comment]
> Hi @forever-rz , Thanks for your interests. It seems that the second codebook is used for quantization while training transformer. For FFHQ, it will not take so long to get a reasonable inpainting results.
Thanks for the reply, but I've carefully compared the posted model to my model training process and really don't notice any difference. So was wondering if it could be the parameters? In the training phase of pvqvae keep_ratio:[0.0,0.5] while the transformer training phase keep_ratio:[0.3,0.6] causes this?

**Comment by liuqk3** [comment]
Hi, @forever-rz . Sorry for the delayed reply.

`keep_ratio` just affects the number of remained pixels in an image, it should not cause such as artifacts. After have a loot at your configs, I do not find something wrong. Here is my questions or suggestions:
1) Did you use our provided P-VQVAE or train it by yourself?
2) Have you checked the reconstruction capability of the used P-VQVAE?
3) Can you provide the cross-entropy loss curves of transformer?

**Comment by forever-rz** [comment]
It seems to me that the reconstruction results are not bad, so I don't understand why transformer's training results are so wrong. Although I made some changes to P-VQVAE, transformer has not changed at all.

**Comment by liuqk3** [comment]
@forever-rz ，
I do not know the number of epochs you have trained on FFHQ and Places2. You can try to visualize the inpainting results of the trained model.

**Comment by myhansu** [comment]
Hi, do you have any updates on this issue? I have also encountered same problem with a custom dataset, reconstruction results are much better than completed results.

**Comment by UESTC-Med424-JYX** [comment]
I also encountered the problem of good reconstruction effect but poor generation effect in a similar task. I saw that our loss curves are basically the same. Did you solve this problem in the end?
